src/builtin/function/math.rs

```diff
@@ -0,0 +1,152 @@
+use serde_json::{Number, Value};
+
+use crate::context::Context;
+use crate::error::{Error, Result};
+
+fn validate_args_len(args: &[Value], expected: usize, name: &'static str) -> Result<()> {
+    if args.len() != expected {
+        return Err(Error::with_message("invalid number of arguments")
+            .context("function", name)
+            .context("argument count", format!("{}", args.len()))
+            .context("expected", format!("{}", expected)));
+    }
+    Ok(())
+}
+
+fn get_one_arg<T, F>(args: &[Value], f: F) -> Result<T>
+where
+    F: FnOnce(&Value) -> Option<T>,
+{
+    args.first()
+        .and_then(f)
+        .ok_or_else(|| Error::with_message("invalid argument type"))
+}
+
+fn get_two_args<T1, T2, F1, F2>(args: &[Value], f1: F1, f2: F2) -> Result<(T1, T2)>
+where
+    F1: FnOnce(&Value) -> Option<T1>,
+    F2: FnOnce(&Value) -> Option<T2>,
+{
+    let r1 = args
+        .get(0)
+        .and_then(f1)
+        .ok_or_else(|| Error::with_message("invalid argument type"))?;
+    let r2 = args
+        .get(1)
+        .and_then(f2)
+        .ok_or_else(|| Error::with_message("invalid argument type"))?;
+    Ok((r1, r2))
+}
+
+pub(crate) fn pow(args: &[Value], _context: &mut Context) -> Result<Value> {
+    validate_args_len(args, 2, "POW")?;
+
+    if let Ok((b, e)) = get_two_args(args, Value::as_i64, Value::as_u64) {
+        return Ok(Value::from(b.pow(e as u32)));
+    }
+
+    let (b, e) = get_two_args(args, Value::as_f64, Value::as_f64)?;
+    Ok(Value::Number(Number::from_f64(b.powf(e)).ok_or_else(|| {
+        Error::with_message("expressions results to NaN").context("function", "POW")
+    })?))
+}
+
+pub(crate) fn log10(args: &[Value], _context: &mut Context) -> Result<Value> {
+    validate_args_len(args, 1, "LOG10")?;
+
+    let x = get_one_arg(args, Value::as_f64)?;
+    Ok(Value::Number(Number::from_f64(x.log10()).ok_or_else(|| {
+        Error::with_message("expressions results to NaN").context("function", "LOG10")
+    })?))
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
These are good example-based tests, I think in the future they could really benefit from being generalized to property-based, using `proptest` for example.